pkg/registry/core/service/storage/rest.go

```diff
@@ -876,3 +890,80 @@ func collectServiceNodePorts(service *api.Service) []int {
 	}
 	return servicePorts
 }
+
+func (rs *REST) getNetworkServiceIPAlloc(service *api.Service) (ipallocator.Interface, error) {
+	netName, ok := service.Labels[arktosv1.NetworkLabel]
+	if !ok {
+		netName = arktosv1.NetworkDefault
+	}
+
+	network, err := rs.getNetwork(genericapirequest.NewContext(), service.Tenant, netName)
+	if err != nil {
+		// todo: remove the temporary measure when we have all the multi-tenant networking components in place
+		// as temporary measure, ignore the error of default network not found
+		if netName == arktosv1.NetworkDefault {
+			return rs.serviceIPs, nil
+		}
+
+		return nil, fmt.Errorf("cannot find network %s of tenant %s: %v", netName, service.Tenant, err)
+	}
+
+	networkUnstructured := network.(*unstructured.Unstructured)
+	var cidrs []string
+	cidrs, ok, err = unstructured.NestedStringSlice(networkUnstructured.Object, "spec", "service", "cidrs")
+	if err != nil {
```

Comment:
I remember in our last discussion we mentioned service CIDR probably isn't required for flat network objects.  Will these check fail for flat network?

Reply:
if flat network does not turn on the feature gate of per-network-service-ip-alloc (the code is in previors part 1 PR), it won't fail. 
this PR is all about turning on the feature, and it will alloc service ip from per-network pool.